DIPs/DIP1xxx-mvf.md

```diff
@@ -0,0 +1,605 @@
+
+### Special case 3
+
+If `e1` is a type, and `prop` is a `static` property of that type, the semantics of `e1` are quite different (i.e. *Requirement 2* and *Requirement 3* are irrelevant). Therefore, it is handled differently.  [Special case 1](#Special%20case%201) and [special case 2](#Special%20case%202) still apply.
+
+```D
+struct S
+{
+    static int field;
+
+    static int prop() @property { return field; }
+    static int prop(int value) @property { return field = value; }
+}
+
+void main()
+{
+    // The lowering of `s.prop += 1` as proposed in this DIP.
+    // Existing Implementation: Compiles and produces the expected result.
+    S.prop(S.prop() + 1);
+    assert(S.field == 1);
+}
+```
+
+#### Module-level properties without a context argument
+
+Module-level properties without a context argument are similar to `static` properties of an aggregate, so they are lowered in a similar manner.   [Special case 1](#Special%20case%201) and [special case 2](#Special%20case%202) still apply.
+
+```D
+int field;
+
+int prop() @property { return field; }
+int prop(int value) @property { return field = value; }
+
+void main()
+{
+    // The lowering of `prop += 1` as proposed in this DIP.
+    // Existing Implementation: Compiles and produces the expected result.
+    prop(prop() + 1);
+    assert(field == 1);
+}
+```
+
+### Free, nested property functions
+
+Nested functions cannot be overloaded, so it is currently not possible to create a getter/setter pair of nested `@property` functions.  Therefore, this implementation will have no effect on nested `@property` functions, maintaining the status quo.
+
+```D
+void main()
+{
+    int mP;
+    int p() @property { return mP; }
+    void p(int v) @property { mP = v; }  // Existing Implementation: Error: declaration p is already defined
+}
+```
+
+### Alternatives
+
+#### Prowiki article
+
+[The Prowiki article](http://www.prowiki.org/wiki4d/wiki.cgi?DocComments/Property#Semanticrewritingofproperties) gave a thorough discussion and many examples, but did not provide any generalization and was too difficult for the author of this DIP to derive an appropriate algorithm from.  However, the article did provide insight that was incorporated into this DIP.
+
+#### DIP4
+
+[DIP4 proposes](https://wiki.dlang.org/DIP4#Usage)...
+
+> `f.width += 5; // Calls getter, saves it in a temp, does += on temp, calls setter with temp`
+
+That translates to:
+
+```D
+auto temp = f.width;
+temp += 5;
+f.width = temp;
+```
+
+That proposal violates Requirements 2.
```

Comment:
Why is it a requirement that `f` is only referenced once?

Reply:
Read the comments starting [here in the implementation](https://github.com/dlang/dmd/pull/7079#issuecomment-327924953) for an explanation.

Reply:
I see. I believe there is a lowering that will reference once and use the correct operator. Maybe :p I'll play around a bit and see.

Reply:
Here, one evaluation, and correct operator. Are there any problems here?

```
import std.stdio : writeln;

struct B {
    this(this) {
        "copy B".writeln;
    }
    void opBinary(string op)(auto ref B other) { writeln("binary: ", op); }
    void opOpAssign(string op)(auto ref B other) { writeln("opassign: ", op); }
}
struct A {
    B b_;
    this(this) {
        "copy A".writeln;
    }
    @property B prop() { "getter".writeln; return b_; }
    @property void prop()(auto ref B b) { "setter".writeln; b_ = b; }
}

void main() {
    A a;
    B b;
    auto array = [A(), A(), A()];
    auto i = 0;
    lower!("prop", "+=")(array[i++], b);
    i.writeln;
}

void lower(string prop, string op, T, U)(auto ref T lhs, ref U rhs) {
    mixin("auto p = lhs." ~ prop ~ ";");
    mixin("p" ~ op ~ "rhs;");
    mixin("lhs." ~ prop ~ "= p;");
}
```